Alignment/APEEstimation/plugins/ApeTreeCreateDefault.cc

```diff
@@ -0,0 +1,473 @@
+  sectorNameTree = new TTree("nameTree","Tree with names of sectors");
+  
+    
+  // Assign the information stored in the trees to arrays
+  std::vector<double*> a_defaultSectorX;
+  std::vector<double*> a_defaultSectorY;
+  
+  std::vector<std::string*> a_sectorName;
+  std::map<unsigned int, TrackerSectorStruct>::const_iterator i_sector;
+  for(i_sector = m_tkSector_.begin(); i_sector != m_tkSector_.end(); ++i_sector){
+    const unsigned int iSector(i_sector->first);
+    const bool pixelSector(i_sector->second.isPixel);
+    
+    a_defaultSectorX.push_back(new double(-99.));
+    a_defaultSectorY.push_back(new double(-99.));
+    a_sectorName.push_back(new std::string(i_sector->second.name));
+    
+    std::stringstream ss_sector;
+    std::stringstream ss_sectorSuffixed;
+    ss_sector << "Ape_Sector_" << iSector;
+  
+    ss_sectorSuffixed << ss_sector.str() << "/D";
+    defaultTreeX->Branch(ss_sector.str().c_str(), &(*a_defaultSectorX[iSector-1]), ss_sectorSuffixed.str().c_str());
+    
+    if(pixelSector){
+      defaultTreeY->Branch(ss_sector.str().c_str(), &(*a_defaultSectorY[iSector-1]), ss_sectorSuffixed.str().c_str());
+    }
+    sectorNameTree->Branch(ss_sector.str().c_str(), &(*a_sectorName[iSector-1]), 32000, 00);  
+  }
+
+  
+  // Loop over sectors for getting default APE
+  
+  for(std::map<unsigned int,TrackerSectorStruct>::iterator i_sector = m_tkSector_.begin(); i_sector != m_tkSector_.end(); ++i_sector){
+    
+    double defaultApeX(0.);
+    double defaultApeY(0.);
+    unsigned int nModules(0);
+    std::vector<unsigned int>::const_iterator i_rawId;
+    for(i_rawId = (*i_sector).second.v_rawId.begin(); i_rawId != (*i_sector).second.v_rawId.end(); ++i_rawId){
+      std::vector<AlignTransformErrorExtended> alignErrors = alignmentErrors->m_alignError;
+      for(std::vector<AlignTransformErrorExtended>::const_iterator i_alignError = alignErrors.begin(); i_alignError != alignErrors.end(); ++i_alignError){
+        if(*i_rawId ==  i_alignError->rawId()){
+          CLHEP::HepSymMatrix errMatrix = i_alignError->matrix();
+          defaultApeX += errMatrix[0][0];
+          defaultApeY += errMatrix[1][1];
+          nModules++;
+        }
+      }
+  }
+  *a_defaultSectorX[(*i_sector).first-1] = defaultApeX/nModules;
+  *a_defaultSectorY[(*i_sector).first-1] = defaultApeY/nModules;
+  }
+  
+  
+  sectorNameTree->Fill();
+  sectorNameTree->Write("nameTree");
+  defaultTreeX->Fill();
+  defaultTreeX->Write("iterTreeX");
+  defaultTreeY->Fill();
+  defaultTreeY->Write("iterTreeY");
+  
+  defaultFile->Close(); 
+}
+  
+
+
+
+// ------------ method called once each job just before starting event loop  ------------
+void 
+ApeTreeCreateDefault::beginJob()
+{
+    
+  this->sectorBuilder();
+}
+
+// ------------ method called once each job just after ending the event loop  ------------
+void 
+ApeTreeCreateDefault::endJob() 
+{
+     
+}
+
+//define this as a plug-in
+DEFINE_FWK_MODULE(ApeTreeCreateDefault);
```

Comment:
in principle one should also add a `fillDescriptions` method that defines the default configuration and gives the framework a handle to validate the configurations

Reply:
If I add a fillDescriptions method, do I have to add a complete default definition of sectors? The sector definition does not really have a simple standard case. It seems like that if I have a vector of ParameterSet, I have to add a default argument.

Reply:
@wddgit could you comment on how this could be done?

Reply:
The documentation is here:

https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideConfigurationValidationAndHelp#ParameterSets_that_Contain_Vecto

and the next two sections about Optional Parameters and Parameters Without Default values are also relevant.

It is possible to call the addVPSet function and omit the last argument, which is the default VPSet. In this case, validation will throw an exception if the VPSet is not explicitly specified in the configuration. And the VPSet will not be written into the generated cfi file at all.

I am not sure I completely understood what you are trying to do. Please feel free to ask more specific questions. I know there are a lot of options in the documentation and it is sometimes hard to find things. And the documentation may not answer your question or leave you confused. I am happy to help more if I understand exactly what you are trying to do.